File: benchmarks/pandas/bench_select_dtypes_options.py
"""Benchmark: DataFrame.select_dtypes() — filter columns by dtype (include/exclude)."""
import json, time
import pandas as pd
import numpy as np

ROWS = 50_000
WARMUP = 5
ITERATIONS = 30

rng = np.random.default_rng(42)
df = pd.DataFrame({
    "intCol": np.arange(ROWS, dtype=np.int32),
    "floatCol": np.arange(ROWS, dtype=np.float64) * 1.5,
    "boolCol": np.arange(ROWS) % 2 == 0,
    "strCol": [f"s_{i % 100}" for i in range(ROWS)],
})

for _ in range(WARMUP):
    df.select_dtypes(include="number")
    df.select_dtypes(exclude="number")
    df.select_dtypes(include=["int", "float"])

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    df.select_dtypes(include="number")
    df.select_dtypes(exclude="number")
    df.select_dtypes(include=["int", "float"])
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({"function": "select_dtypes_options", "mean_ms": round(total_ms / ITERATIONS, 3), "iterations": ITERATIONS, "total_ms": round(total_ms, 3)}))

File: benchmarks/tsb/bench_select_dtypes_options.ts
/**
 * Benchmark: selectDtypes — filter DataFrame columns by dtype (include/exclude).
 * Outputs JSON: {"function": "select_dtypes_options", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { DataFrame, Series, selectDtypes } from "../../src/index.ts";

const ROWS = 50_000;
const WARMUP = 5;
const ITERATIONS = 30;

// Build a mixed-dtype DataFrame
const intCol = new Series({ data: Int32Array.from({ length: ROWS }, (_, i) => i) });
const floatCol = new Series({ data: Float64Array.from({ length: ROWS }, (_, i) => i * 1.5) });
const boolCol = new Series({ data: Array.from({ length: ROWS }, (_, i) => i % 2 === 0) });
const strCol = new Series({ data: Array.from({ length: ROWS }, (_, i) => `s_${i % 100}`) });
const df = DataFrame.fromColumns({ intCol, floatCol, boolCol, strCol });

for (let i = 0; i < WARMUP; i++) {
  selectDtypes(df, { include: "number" });
  selectDtypes(df, { exclude: "number" });
  selectDtypes(df, { include: ["integer", "float"] });
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const start = performance.now();
  selectDtypes(df, { include: "number" });
  selectDtypes(df, { exclude: "number" });
  selectDtypes(df, { include: ["integer", "float"] });
  times.push(performance.now() - start);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "select_dtypes_options",
    mean_ms: Math.round(meanMs * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(totalMs * 1000) / 1000,
  }),
);
